File: tests/algorithms/transpiler/FS/ciphers/permutation_cipher.fs
// Generated 2025-08-07 10:31 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

let _idx (arr:'a array) (i:int) : 'a =
    if i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    let s = sprintf "%A" v
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let mutable seed: int = 1
let rec rand (max: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable max = max
    try
        seed <- ((((seed * 1103515245) + 12345) % 2147483647 + 2147483647) % 2147483647)
        __ret <- ((seed % max + max) % max)
        raise Return
        __ret
    with
        | Return -> __ret
let rec generate_valid_block_size (message_length: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable message_length = message_length
    try
        let mutable factors: int array = [||]
        let mutable i: int = 2
        while i <= message_length do
            if (((message_length % i + i) % i)) = 0 then
                factors <- Array.append factors [|i|]
            i <- i + 1
        let idx: int = rand (Seq.length (factors))
        __ret <- _idx factors (idx)
        raise Return
        __ret
    with
        | Return -> __ret
let rec generate_permutation_key (block_size: int) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable block_size = block_size
    try
        let mutable digits: int array = [||]
        let mutable i: int = 0
        while i < block_size do
            digits <- Array.append digits [|i|]
            i <- i + 1
        let mutable j: int = block_size - 1
        while j > 0 do
            let k: int = rand (j + 1)
            let temp: int = _idx digits (j)
            digits.[j] <- _idx digits (k)
            digits.[k] <- temp
            j <- j - 1
        __ret <- digits
        raise Return
        __ret
    with
        | Return -> __ret
let rec encrypt (message: string) (key: int array) (block_size: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable message = message
    let mutable key = key
    let mutable block_size = block_size
    try
        let mutable encrypted: string = ""
        let mutable i: int = 0
        while i < (String.length (message)) do
            let block: string = _substring message i (i + block_size)
            let mutable j: int = 0
            while j < block_size do
                encrypted <- encrypted + (_substring block (_idx key (j)) ((_idx key (j)) + 1))
                j <- j + 1
            i <- i + block_size
        __ret <- encrypted
        raise Return
        __ret
    with
        | Return -> __ret
let rec repeat_string (times: int) =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    let mutable times = times
    try
        let mutable res: string array = [||]
        let mutable i: int = 0
        while i < times do
            res <- Array.append res [|""|]
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let rec decrypt (encrypted: string) (key: int array) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable encrypted = encrypted
    let mutable key = key
    try
        let klen: int = Seq.length (key)
        let mutable decrypted: string = ""
        let mutable i: int = 0
        while i < (String.length (encrypted)) do
            let block: string = _substring encrypted i (i + klen)
            let mutable original: string array = repeat_string (klen)
            let mutable j: int = 0
            while j < klen do
                original.[_idx key (j)] <- _substring block j (j + 1)
                j <- j + 1
            j <- 0
            while j < klen do
                decrypted <- decrypted + (_idx original (j))
                j <- j + 1
            i <- i + klen
        __ret <- decrypted
        raise Return
        __ret
    with
        | Return -> __ret
let message: string = "HELLO WORLD"
let block_size: int = generate_valid_block_size (String.length (message))
let key: int array = generate_permutation_key (block_size)
let mutable encrypted: string = encrypt (message) (key) (block_size)
let mutable decrypted: string = decrypt (encrypted) (key)
printfn "%s" ("Block size: " + (_str (block_size)))
printfn "%s" ("Key: " + (_str (key)))
printfn "%s" ("Encrypted: " + encrypted)
printfn "%s" ("Decrypted: " + decrypted)
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/algorithms/x/PHP/ciphers/permutation_cipher.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$__start_mem = memory_get_usage();
$__start = _now();
  $seed = 1;
  function mochi_rand($max) {
  global $seed, $message, $block_size, $key, $encrypted, $decrypted;
  $seed = ($seed * 1103515245 + 12345) % 2147483647;
  return $seed % $max;
};
  function generate_valid_block_size($message_length) {
  global $seed, $message, $block_size, $key, $encrypted, $decrypted;
  $factors = [];
  $i = 2;
  while ($i <= $message_length) {
  if ($message_length % $i == 0) {
  $factors = _append($factors, $i);
}
  $i = $i + 1;
};
  $idx = mochi_rand(count($factors));
  return $factors[$idx];
};
  function generate_permutation_key($block_size) {
  global $seed, $message, $key, $encrypted, $decrypted;
  $digits = [];
  $i = 0;
  while ($i < $block_size) {
  $digits = _append($digits, $i);
  $i = $i + 1;
};
  $j = $block_size - 1;
  while ($j > 0) {
  $k = mochi_rand($j + 1);
  $temp = $digits[$j];
  $digits[$j] = $digits[$k];
  $digits[$k] = $temp;
  $j = $j - 1;
};
  return $digits;
};
  function encrypt($message, $key, $block_size) {
  global $seed, $decrypted;
  $encrypted = '';
  $i = 0;
  while ($i < strlen($message)) {
  $block = substr($message, $i, $i + $block_size - $i);
  $j = 0;
  while ($j < $block_size) {
  $encrypted = $encrypted . substr($block, $key[$j], $key[$j] + 1 - $key[$j]);
  $j = $j + 1;
};
  $i = $i + $block_size;
};
  return $encrypted;
};
  function repeat_string($times) {
  global $seed, $message, $block_size, $key, $encrypted, $decrypted;
  $res = [];
  $i = 0;
  while ($i < $times) {
  $res = _append($res, '');
  $i = $i + 1;
};
  return $res;
};
  function decrypt($encrypted, $key) {
  global $seed, $message, $block_size;
  $klen = count($key);
  $decrypted = '';
  $i = 0;
  while ($i < strlen($encrypted)) {
  $block = substr($encrypted, $i, $i + $klen - $i);
  $original = repeat_string($klen);
  $j = 0;
  while ($j < $klen) {
  $original[$key[$j]] = substr($block, $j, $j + 1 - $j);
  $j = $j + 1;
};
  $j = 0;
  while ($j < $klen) {
  $decrypted = $decrypted . $original[$j];
  $j = $j + 1;
};
  $i = $i + $klen;
};
  return $decrypted;
};
  $message = 'HELLO WORLD';
  $block_size = generate_valid_block_size(strlen($message));
  $key = generate_permutation_key($block_size);
  $encrypted = encrypt($message, $key, $block_size);
  $decrypted = decrypt($encrypted, $key);
  echo rtrim('Block size: ' . _str($block_size)), PHP_EOL;
  echo rtrim('Key: ' . _str($key)), PHP_EOL;
  echo rtrim('Encrypted: ' . $encrypted), PHP_EOL;
  echo rtrim('Decrypted: ' . $decrypted), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
